### ISSUE TITLE
Get wi/cs from release diff instead of build diff

### ISSUE THREAD

**Issue by jonathanMelly** [rfennell/AzurePipelines#695]
### Azure DevOps Extension you are using
Generate Release Notes (Crossplatform) 2.38.2 (public version)

### Where are you running it?
- Azure DevOps Service (VSTS)

### Expected behaviour and actual behaviour
I would like that WI/CS could be retrieved from a **release diff** ([as offered by release api](https://vsrm.dev.azure.com/jonathanmelly/TestIntellijAndRelease/_apis/Release/releases/7/workitems?baseReleaseId=6)) instead of a **[build diff](https://dev.azure.com/jonathanMelly/TestIntellijAndRelease/_apis/build/workitems?fromBuildId=65&toBuildId=68&api-version=5.1-preview.2)**.

As I am using a github repository, wi/cs are not automatically linked upon build ([feature requested](https://developercommunity.visualstudio.com/content/idea/1009927/automatically-link-work-items-for-builds-done-with.html)), but the release API is capable of doing the job thus I would be glad to have this option (a new setting for the task)...

### Steps to reproduce the problem

1. Create a build pipeline based on a github repo
2. Commit and push stuff with "AB#XX" references in commit messages
3. Create a release based on build artifact with release notes generation
4. There is no wi/cs found

Would you agree to integrate that feature if I submit a PR or do you know another way to make this work ?

**Comment by rfennell** [comment]
Sounds in interesting idea and always happy to consider PRs.

What I would say is that any enhancement such as this should not introduce breaking changes, so I would like to see the new feature only enabled by setting some boolean task parameter, and the default behaviour remaining as it is now.

If you don't have time to look into it, I will get around to investigating the enhancement, just not sure when

**Comment by jonathanMelly** [comment]
Thanks, doing it as you suggest (boolean flag to enable separated new code / let as is) sounds even more easy, thus the biggest part of the work for me submitting a PR will be to setup a dev environment... which sounds quite interesting ;-)
I hope to submit a PR next week if you haven't already done so by then.

**Comment by jonathanMelly** [comment]
Hello, I just wanna share [a first draft](https://github.com/jonathanMelly/AzurePipelines/commit/7b9447aa22168f59c1a0f6ec3ab3d9bfc461a8dc) mainly to tell you that I'm working on this and that I expect to submit the PR next week (tests and refactoring to do).

**Comment by rfennell** [comment]
Looks good, just to make you aware I have just update this extension to move from the old VSO named Node API to the newer Azure-DevOps one. This was to try to address issue seen in #694 due to the old API inferring Team Project IDs. I thought it best to move to the current version which requires them to be passed into all methods that requirement.

This may cause you some merge issues, sorry

**Comment by jonathanMelly** [comment]
Ok no problem for the merge :-)
While testing the code I added, I get this error:
![security error](https://storage.prod.mdl.swisscom.ch/object/~S0015F4A1F98D3E437CC87F9892BF61CC56E56709A5?p=ZXJyb3IucG5n)
Do you have any clue on this ?

**Comment by rfennell** [comment]
Not seen that before, what does it say before in the light, running with the variable system.debug=true gives more detail.

**Comment by jonathanMelly** [comment]
I tried to add some random 'admin rights' to the agent without success...
Here is a log extract  with debug ([full log here](https://storage.prod.mdl.swisscom.ch/object/~S00BC05861F276A78B08F415FED683E40B3281C796F?p=dGFza2xvZ182LmxvZw)):
2020-05-01T16:10:01.9338523Z Is Primary: [true]
2020-05-01T16:10:01.9338910Z Looking for the [_publish] in the most recent successful release [Release-14]
2020-05-01T16:10:01.9339410Z Found artifact [_publish] with build number [20200501.6] in release [Release-14]
2020-05-01T16:10:01.9339922Z Checking what commits and workitems have changed from [20200501.6][ID 79] => [20200501.7] [ID 80]
2020-05-01T16:10:01.9340620Z ##[debug]ReleaseNotes.Fix349=undefined
2020-05-01T16:10:01.9341118Z Defaulting on the workaround for build API limitation (see issue #349 set 'ReleaseNotes.Fix349=false' to disable)
2020-05-01T16:10:01.9341637Z Using workaround for build API limitation (see issue #349)
2020-05-01T16:10:02.2066022Z Detected 0 commits/changesets and 0 workitems between the builds.
2020-05-01T16:10:02.2066991Z 
2020-05-01T16:10:02.2081762Z ##[debug]alsoGatherDataFromReleaseApi=true
2020-05-01T16:10:02.2082820Z Looking for WI/CS between release 16 and 17
2020-05-01T16:10:02.3282136Z ##[error]Cannot retrieve WI from release api : Error: You do not have permissions to perform this operation on the service connection. An Endpoint Administrator should add you to the Endpoint Readers group of this service connection.
2020-05-01T16:10:02.3293229Z ##[debug]Processed: ##vso[task.issue type=error;]Cannot retrieve WI from release api : Error: You do not have permissions to perform this operation on the service connection. An Endpoint Administrator should add you to the Endpoint Readers group of this service connection.
2020-05-01T16:10:02.4469582Z ##[error]Cannot retrieve WS from release api : Error: You do not have permissions to perform this operation on the service connection. An Endpoint Administrator should add you to the Endpoint Readers group of this service connection.
2020-05-01T16:10:02.4470996Z ##[debug]Processed: ##vso[task.issue type=error;]Cannot retrieve WS from release api : Error: You do not have permissions to perform this operation on the service connection. An Endpoint Administrator should add you to the Endpoint Readers group of this service connection.
2020-05-01T16:10:02.4472258Z Detected 0 commits/changesets and 0 workitems between the releases 16 and 17.
2020-05-01T16:10:02.4476391Z Get details of [0] WIs
2020-05-01T16:10:02.4476942Z Total build artifacts: [1
2020-05-01T16:10:02.4477268Z Total commits: [0]
2020-05-01T16:10:02.4477597Z Total workitems: [0]
2020-05-01T16:10:02.4478153Z Leaving WI in default order as returned by API
2020-05-01T16:10:02.4478641Z ##[debug]Build.BuildId=80

The error doesn't come when there are no results...which makes me realize that I never tested the standard extension with data (as I have no ref to builds...) and maybe the issue would be the same...
I'll merge with new endpoints and give another try and further investigate if needed... Thanks a lot for your help even if you don't have the answer :-).

**Comment by rfennell** [comment]
Remember that the task requires that the 'use OAUTH' is set for the agent running in a release, which I assume you are to be calling the release API. Could that be the issue?

**Comment by jonathanMelly** [comment]
Unfortunately not (oauth token option activated).
Merging with new 'azure-devops' namespace doesn't help either...
I guess the token doesn't allow 'full access' on the api and am trying to find a way to authorize that...

**Comment by rfennell** [comment]
The token means you are running as the 'project xxx build service' or similar you can grant that users any rights. You have have to grant special right as detailed in the wiki for for my [build update task](https://github.com/rfennell/AzurePipelines/wiki/BuildTasks-Task). I wonder if helps

I  had a think about this feature, not to say it is not valuable now, but I wonder how much releases will matter as we move towards multi stage YAML as I have just fone for all these tasks, [see video](https://youtu.be/WMQ0G9eXczE)

**Comment by jonathanMelly** [comment]
Thanks I tried that yesterday without success when modifying Manage releases => Allow... I may try with Edit release stage/definition as it looks like 'edit build definition'...

**Comment by jonathanMelly** [comment]
Hi again, I granted all accesses to the build service without success, thus the only solution I know is to use a PAT.

Concerning the multi-stage YAML it won't help my scenario as the problem is that azureDevops platform is not (yet?) able to link work items referenced on github commits (AB#XXX) upon build but do it correctly when comparing releases (reason for this feature) ?!?.

From my point of view, I see 2 possibilities:
1) I keep this feature for private use (issue can be closed)
2) You agree to integrate this feature knowing that it requires a PAT

What do you think ?

**Comment by rfennell** [comment]
I am surprised that there is not a way to grant the service user access to the endpoint, I an going have a look into that as I am intrigued.

Assuming we cannot find a way around the service account issues, as to the route forward I think I would favour creating a new task to use your logic, taking the chance to gut out anything that is not required such as the old legacy template model. This could be published privately or publically depending on the appetite for support question.

The danger of placing it inside this task will a special PAT is that it is confusing. That said I have just had to add a special PAT for expanding commit message for commits on private GitHub repos as the PAT used to clone the repo is not exposed to the tasks.

I will look into that service account issue

**Comment by jonathanMelly** [comment]
Ok thank you for the feedback.

Meanwhile I did more tests with multistage-yaml without much success with wi references...

I would also be curious about the reason for the security issue :-)

Thanks to your extension and your help I now have a working scenario to generate a github release based on azure devops work items. I am surprised to be the only one in this situation (using azure devops planning features and github repo and willing to have an efficient release mechanism based on available data) or maybe people just duplicate their repo...

Anyway thanks again, I will keep a fork and privately deploy the adapted version for my context.

As you may get some answers regarding the security issue I do not close this issue yet (while understanding if you do it)

**Comment by rfennell** [comment]
Strange, my initial tests don't show the same problem with access you have. I just hacked  the JS file with this as the first thing run when it is release using the task 

```
agentApi.logInfo("before test");
let xxx = yield releaseApi.getReleaseWorkItemsRefs(teamProject, 2654, 2653);
agentApi.logInfo(xxx.length);
agentApi.logInfo("after test");
```
>  used the technique of using a private agent, run a release so the task get cached locally in `<agent folder>/_work/_tasks/xplat....` . You can than replace the generatereleasenotes.js with a new version, or just edit the Javascript. Avoid a long deploy cycle - I use it all the time in debugging

This returns 6 items as I would expect.

Stranger and stranger.

Run out of time today, but I will keep digging

**Comment by jonathanMelly** [comment]
Thanks for the news and the tip :smile:

What you have is what I was expecting to get :sweat_smile:

I wonder if it is linked to the setup with a github repo as when there are no wi/cs between the releases, it works as well on my side so the problem would come when the api tries to populate some data of wi/cs which is on github side :confused: ??

I may try again in a brand new project...

**Comment by rfennell** [comment]
I have managed to repo the problem, the key thing was to follow you very clear instructions not 'just go for it' 😄

Now when I make the `.getReleaseWorkItemsRefs()` call if there is an `AB#123` reference I see the same error relating to the Azure Boards/ GitHub connector.

If you look at this connector in the settings, you can see it does not provide an option to edit the security.

![image](https://user-images.githubusercontent.com/324718/80967554-74e3e480-8e0e-11ea-98e5-26db4e8f0f66.png)

This I think is the root cause. I will ask some contacts at MS to see if there is an answer other than using a PAT

**Comment by jonathanMelly** [comment]
:astonished:  Thanks a lot for this inquiry :relieved:
I hope that your contacts will be interested in making a little patch :blush:

And I can stop testing on my side :sleepy:

**Comment by rfennell** [comment]
No getting any answer back on this, so I have logged an issue on [Git Repo for the API](https://github.com/microsoft/azure-devops-node-api/issues/394)

**Comment by rfennell** [comment]
The GitHub issue was closed, I was told to log the issue  [here](https://developercommunity.visualstudio.com/content/problem/1025945/cannot-call-getreleaseworkitemsrefs-using-the-pipe.html)

**Comment by jonathanMelly** [comment]
Thanks for your help

**Comment by rfennell** [comment]
@jonathanMelly I have a solution, well a workaround

based on the feedback form Microsoft ​I did more experimenting with https://dev.azure.com/{your_organization}/{your_project}/_settings/adminservices. and I found that if I grant 'user' rights to my '<Project name> build service' to all service connections the problem is resolved. 'Reader' rights are not enough.

As this seemed a bit excessive, I then dug into the exact service connection, setting the 'user' right for the account on each Service Connector with GitHub in it's name, and by process of elimination found the right one.

At this time I am not sure of the way to find the service connection required in a more scientific way.

**Comment by jonathanMelly** [comment]
Thanks Again.
Could you give me more details on how to achieve this because I am not able to reproduce what you describe:
![security](https://storage.prod.mdl.swisscom.ch/thumbnail/~S0035F51D2BD256BEF5DE439F08E9D478085297AA82?p=L3NlY3VyaXR5LnBuZw&height=324)

**Comment by jonathanMelly** [comment]
OK it was just that the search feature is not as smart as I was expecting.
To find the correct endpoint, I used the 'usage history' tab.

**Comment by rfennell** [comment]
Strange I have got no useful information in the usage history, also not sure why in my case it is picking the service connection it is. But glad you are unblocked for now

**Comment by jonathanMelly** [comment]
Yes no more PAT for my build ;-)

Does what you discovered change your mind in keeping this feature in another task or may I close this issue and wait for any prospective demand about this to make a contribution with a PR ?

**Comment by rfennell** [comment]
If you want to send in a PR happy to have a look at it. I guess it won't have much code, but it is going to need good documentation 

So yes might as well close this as we have a workaround

**Comment by jonathanMelly** [comment]
Hi again, as I was testing my PR before submitting it I realized that the behavior is not the same with the security endpoint configuration (your last suggestion) and using the PAT.

With the PAT, wi and cs are found and with the security configuration, only the cs are found (no error reported for the wi) !

Thus the security solution is a nogo for now... as for my PR with beautiful documentation :sweat_smile:

I wonder what is under the hood of this behavior but it's seems to be complicated...

**Comment by rfennell** [comment]
I am continuing to talk to MS to get a better understanding of this behaviour.

I think it might be using multiple calls to different connectors, and in my testing I am getting WI back. So worth more experimentation. 

If you want to fork my repo and create a private task with just your functionality feel free - it might easier support-wise

**Comment by jonathanMelly** [comment]
> I think it might be using multiple calls to different connectors, and in my testing I am getting WI back. So worth more experimentation.

I have authorized my build service to all connectors and still don't get the WI's...

> If you want to fork my repo and create a private task with just your functionality feel free - it might easier support-wise

OK thanks.

Have a nice day ;-)

**Comment by rfennell** [comment]
@jonathanMelly how is this going? do you think it is worth keeping this issue open? Or shall I mark it as being changed in a private fork?

**Comment by jonathanMelly** [comment]
@rfennell Please mark it as being changed in a [private fork](https://github.com/jonathanMelly/AzurePipelines) (as I was not able to make it work with what you discovered on security configuration)

**Comment by rfennell** [comment]
As commented above this proved to be a problematic change due to security limitations. It has been handled as a private fork
